### PR TITLE
ENH: Refactor the model so it is compatible with DIPY's interface

### PR DESCRIPTION
**Pull request by oesteban**
The strategy is to initiate the new `eddymotion.model.dipy`, where we test out DIPY "extensions" which potentially will be added to their codebase. In it, a new GP model is added.

Our base module is then updated to handle the GP model as a standard DIPY model.

cc/ @arokem

### CODE REVIEW

File: src/eddymotion/model/dipy.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#     https://www.nipreps.org/community/licensing/
+#
+"""DIPY-like models (a sandbox to trial them out before upstreaming to DIPY)."""
+
+# import numpy as np
+
+
+from sklearn.gaussian_process import GaussianProcessRegressor
+from dipy.reconst.base import ReconstModel
+
```
Comment (arokem) [inline src/eddymotion/model/dipy.py:47]:
```suggestion
from dipy.reconst.multi_voxel import multi_voxel_fit
```

File: src/eddymotion/model/dipy.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+           doi:\
+           `10.1016/j.neuroimage.2015.07.067 <https://doi.org/10.1016/j.neuroimage.2015.07.067>`__.
+
+        """
+
+        ReconstModel.__init__(self, gtab)
+        self.kernel_model = kernel_model
+        self.mask = mask
+
+    def fit(self, data, mask=None, random_state=0):
```
Comment (arokem) [inline src/eddymotion/model/dipy.py]:
```suggestion
    @multi_voxel_fit
    def fit(self, data, mask=None, random_state=0, **kwargs):
```

Comment (arokem) [reply src/eddymotion/model/dipy.py]: The additional kwargs let you control the parallelization.

**Review by arokem — COMMENTED** [review]
Looking good. This is how you implement parallelization as in https://github.com/dipy/dipy/pull/2593

File: src/eddymotion/model/dipy.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+        gtab = gtab if gtab is not None else self.gtab
+        return GPFit(
+            self._gpr.fit(gtab, data),
+            gtab=gtab,
+            mask=mask,
+        )
+
+    @multi_voxel_fit
+    def multi_fit(self, data_thres, mask=None, **kwargs):
```
Comment (arokem) [inline src/eddymotion/model/dipy.py]: What is the purpose of this method? Also, looks like the mask doesn't get used below.

**Review by arokem — COMMENTED** [review]
This is great! With a bit more unit testing, I think that this will be easy to port upstream to DIPY down the line. 

For now, I am just wondering about the separation between `fit` and `multi_fit`. What's the purpose of that?

File: src/eddymotion/model/dipy.py (hunk 3 above)
Comment (oesteban) [reply src/eddymotion/model/dipy.py]: I separated the decorated fit because if I interpret the `multi_voxel_fit` decorator properly, an array with shape (n_parameters, n_voxels) is expected as the output of the decorated function, so that then parameters can be stacked/manipulated and used in prediction.

In our case, instead, we get a trained GaussianProcessRegressor object instance. We can probably extract the parameters that describe it, but I don't know how to do that at this point.

So I moved it out to punt the implementation at a later time when everything is working for a single thread single voxel fit.

WDTY?

File: src/eddymotion/model/dipy.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    """
+
+    # Check it's fitted as they do in sklearn internally
+    # https://github.com/scikit-learn/scikit-learn/blob/972e17fe1aa12d481b120ad4a3dc076bae736931/\
+    # sklearn/gaussian_process/_gpr.py#L410C9-L410C42
+    if not hasattr(model._gpr, "X_train_"):
+        raise RuntimeError("Model is not yet fitted.")
+
+    # Extract orientations from gtab, and highly likely, the b-value too.
+    return model._gpr.predict(gtab, return_std=False)
```
Comment (oesteban) [inline src/eddymotion/model/dipy.py:71]: @jhlegarreta I'm going to walk back my words and now I'm confident that the reshaping (w/ or w/o mask) should happen at a higher level of the class hierarchy. Basically, all models we have so far need to do it, so let's not re-implement the same operation for every model.

WDTY?

**Comment by oesteban** [comment]
@jhlegarreta are you comfortable merging this now? That will allow me to make further steps before tomorrow's meeting.

I believe with my last commit, things are in a very good state to be set up as a base for the GP.

File: src/eddymotion/model/dipy.py (hunk 4 above)
Comment (jhlegarreta) [reply src/eddymotion/model/dipy.py:71]: Seems reasonable. Did not have time to have a look yesterday, so I am probably missing some context, but if you now have a clearer picture across all models and believe so, I am fine with it.

**Comment by jhlegarreta** [comment]
> @jhlegarreta are you comfortable merging this now? That will allow me to make further steps before tomorrow's meeting.

A few comments:
- I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository: https://github.com/nipreps/eddymotion/pull/212.
- The intermediate commits (removing the `viz` module, adding the covariance models, etc.) make the review  more difficult. Rebasing on the fork `master` should have been enough to only leave the relevant commits and changes.
- Moving things (gradient angle computation, covariance computation, etc.) to a new `.dipy` module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

But I guess my comments are of little value if this needs to be merged now. @oesteban let me know if you really believe that addressing my comments is unfeasible/not useful, and will merge.

**Comment by oesteban** [comment]
> * I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository:

My bad, I missed that one. Will work on it tomorrow, merge and then downstream here.

> The intermediate commits (removing the viz module, adding the covariance models, etc.) make the review more difficult. Rebasing on the fork master should have been enough to only leave the relevant commits and changes.

I merged upstream/main into my branch precisely to make it easier to review, but yes, your branch would require a rebase to make the diff easier.

**Comment by oesteban** [comment]
Failed to answer the last one:

> Moving things (gradient angle computation, covariance computation, etc.) to a new .dipy module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

Should this new model be upstreamed to DIPY, these functions must also be propagated there or call equivalent and pre-existing implementations. I do not foresee many more models being upstreamed to DIPY, so it feels okay if all these items are kept together without proliferating small modules. Regarding debugging and testing, I'm not so clear that having these functions scattered in small modules is helpful. On the contrary, it may be easier if the stack trace is in a single file.

File: src/eddymotion/model/dipy.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    """
+
+    _ensure_positive_scale(a)
+
+    return np.where(theta <= a, 1 - 3 * (theta / a) ** 2 + 2 * (theta / a) ** 3, 0)
+
+
+def compute_pairwise_angles(
+    gtab_X: GradientTable,
+    gtab_Y: GradientTable | None = None,
```
Comment (oesteban) [inline src/eddymotion/model/dipy.py:354]: Since this will probably make it into dipy one day, my point of having these being ndarrays does not stand. Better use the gradient table object, which gives us two goodies:

- implicit checking of dimensionality (and therefore, dropped checks within the implementation)
- access to the b-values (which will come handy when we support multi-shell kernels).

File: src/eddymotion/model/dipy.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+       correction for off-resonance effects and subject movement in diffusion MR
+       imaging, NeuroImage 125 (2016) 1063–1078
+
+    """
+
+    bvecs_X = gtab_X.bvecs.T
+    bvecs_X = np.array(bvecs_X) / np.linalg.norm(bvecs_X, axis=0)
+
+    if gtab_Y is None:
+        bvecs_Y = bvecs_X
```
Comment (oesteban) [inline src/eddymotion/model/dipy.py:411]: Support the old use case (auto-covariance) if the second vector is not provided.

**Comment by oesteban** [comment]
>     * I do not understand why e.g. the changes to the signature of the angle computation would not be reviewed and eventually merged in the upstream repository: [ENH: Make grad vec angle computation method accept two input parameters nipreps/eddymotion#212](https://github.com/nipreps/eddymotion/pull/212).

Addressed - #212 is now merged and I have updated downstream. Apologies for the oversight.
 
>     * The intermediate commits (removing the `viz` module, adding the covariance models, etc.) make the review  more difficult. Rebasing on the fork `master` should have been enough to only leave the relevant commits and changes.

I have rebased your branch and this PR - diff should be clean now.


>     * Moving things (gradient angle computation, covariance computation, etc.) to a new `.dipy` module that are not strictly related to the fit/predict wrappers or multi-voxel decorators seems a little bit too early/counter-intuitive to me; I do not believe they make debugging the core of the GP thing easier at this point. I believe that keeping them separate as they were in the upstream repository allowed to focus better on the parts that need attention/debugging (and even when the whole thing may be moved to DIPY).

This is the only outstanding point, which I hope I made sense responding above - let's chat later :)

File: src/eddymotion/model/dipy.py (hunk 5 above)
Comment (jhlegarreta) [reply src/eddymotion/model/dipy.py:354]: :+1: Resolving.

File: src/eddymotion/model/dipy.py (hunk 6 above)
Comment (jhlegarreta) [reply src/eddymotion/model/dipy.py:411]: :+1: Resolving. I'll add a test case for this in a separate PR.

**Comment by jhlegarreta** [comment]
OK, I'm merging this after our conversation, and we move forward from there. Thanks Óscar.
